### PR TITLE
Deferred response with response formatter

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️

Alternative to https://github.com/yiisoft/yii-demo/pull/72

### CODE REVIEW

File: src/Controller.php [hunk 1]
```diff
@@ -33,11 +34,12 @@ public function __construct(
...
-        $response = $this->responseFactory->createResponse();
-        $content = $this->view->render($view, $parameters, $this);
-        $response->getBody()->write($this->renderContent($content));
+        $controller = $this;
+        $contentRenderer = function () use ($view, $parameters, $controller) {
+            return $controller->renderContent($controller->view->render($view, $parameters, $controller));
+        };
 
-        return $response;
+        return $this->factory->create(DeferredResponse::class, [$contentRenderer]);
```
Comment (samdark) [inline src/Controller.php]: Likely could be made more explicit about dependencies:

1. Create `DefferedResponseFactory` that creates `DeferredResponse`.
2. Inject it into constructor.

File: src/Factory/AppRouterFactory.php [hunk 2]
```diff
@@ -38,12 +42,21 @@ public function __invoke(ContainerInterface $container)
...
                 Route::get('/{login}', [UserController::class, 'profile'])
                      ->name('user/profile'),
             ]),
 
+            // User
+            Group::create('/api', [
+                Route::get('/user', [ApiUserController::class, 'index'])
+                    ->name('api/user/index'),
+                Route::get('/user/{login}', [ApiUserController::class, 'profile'])
+                    ->addMiddleware(new ResponseFormatter($container->get(JsonResponseFormatter::class)))
```
Comment (samdark) [inline src/Factory/AppRouterFactory.php]: Do we need non-deferred responses at all?

File: src/Controller.php (hunk 1 above)
Comment (yiiliveext) [reply src/Controller.php]: Done

File: .gitignore [hunk 3]
```diff
@@ -25,3 +25,5 @@ phpunit.phar
 /vendor
 /composer.lock
 /node_modules
+
+package.json
```
Comment (rustamwin) [inline .gitignore]: Why?
```suggestion
package-lock.json
```

Comment (yiiliveext) [reply .gitignore]: Scrutinizer generates an error with it. Temporary disabled. This file is regenerated every time on the composer update.

File: src/DeferredResponseFactory.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    public function createResponse(int $code = 200, string $reasonPhrase = '', $data = null): ResponseInterface
+    {
+        return new DeferredResponse($data, $this->createResponseInternal($code, $reasonPhrase), $this->streamFactory);
+    }
+
+    private function createResponseInternal(int $code = 200, string $reasonPhrase = ''): ResponseInterface
+    {
+        if (2 > \func_num_args()) {
```
Comment (samdark) [inline src/DeferredResponseFactory.php]: How does that work? There are exactly two arguments for `createResponseInternal()` passed so the check should never be true.

File: src/HtmlResponseFormatter.php [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    /**
+     * @var string the XML encoding. If not set, it will use the value of [[Response::charset]].
+     */
+    private string $encoding = 'UTF-8';
+
+    public function format(DeferredResponse $deferredResponse): ResponseInterface
+    {
+        $data = $deferredResponse->getData();
+        $response = $deferredResponse->getResponse();
+        $response->getBody()->write($data);
```
Comment (samdark) [inline src/HtmlResponseFormatter.php]: Should we cast `$data` to string here?

File: src/DeferredResponseFactory.php (hunk 4 above)
Comment (yiiliveext) [reply src/DeferredResponseFactory.php]: This a copy of the code from Psr17Factory

Comment (samdark) [reply src/DeferredResponseFactory.php]: I don't understand it.

File: src/HtmlResponseFormatter.php (hunk 5 above)
Comment (yiiliveext) [reply src/HtmlResponseFormatter.php]: Done

**Comment by samdark** [comment]
Do not merge. See https://github.com/yiisoft/yii-web/pull/233

**Pull request merged**